### ISSUE TITLE
Using Riverpod StateNotifier with Navigator 2.0

### ISSUE THREAD

**Issue by echonox** [rrousselGit/riverpod#946]
**Describe what scenario you think is uncovered by the existing examples/articles**
All of the examples on the internet use the ChangeNotifer with the provider package, but what about StateNotifier?
I've already implemented StateNotifier in my Navigator 2.0 and it works fine but I was wondering how would I implement this part of the RouterDelegate:

```@override
  void addListener(VoidCallback listener) {
    // TODO: implement addListener
  }

  @override
  void removeListener(VoidCallback listener) {
    // TODO: implement removeListener
  }
```

As when I don't provide the ChageNotifier mixin (as I am using StateNotifier and not a ChangeNotifier) and Riverpod doesn't seem to provide a mixin, vscode alerts me that I need to add the two snippets above.

So is the implementation above needed in the case of StateNotifier and if yes how would I do that?

**Describe why existing examples/articles do not cover this case**
Explain which examples/articles you have seen before making this request, and
why they did not help you with your problem.

I've searched basically any article and Github issue both here and on the flutter repository.

**Additional context**
Add any other context or screenshots about the documentation request here.

**Comment by daniellampl** [comment]
I currently connect `StateNotifierProvider`s and `StateProvider`s with my `AppRouterDelegate` as follows:

```
final selectedIdProvider = StateProvider<int?>((ref) {
  return null;
});

class AppRouterDelegate extends RouterDelegate<AppRoute> with PopNavigatorRouterDelegateMixin<AppRoute>, ChangeNotifier {
  final WidgetRef ref;

  AppRouterDelegate(this.ref) {
    // listen to provider changes
    ref.listen<int?>(selectedIdProvider, (_, __) => notifyListeners());
  }

  int? get selectedId => ref.read(selectedIdProvider);

  @override
  GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();

  @override
  Future<void> setNewRoutePath(AppRoute configuration) async => SynchronousFuture(null);

  @override
  Widget build(BuildContext context) {
    return Navigator(
      key: navigatorKey,
      pages: [
        const MaterialPage(key: ValueKey('home'), child: HomeView()),
        if (selectedId != null) ...[
          MaterialPage(key: const ValueKey('d'), child: ItemDetailView(itemId: selectedId!)),
        ]
      ],
      onPopPage: (Route<dynamic> route, dynamic result) {
        // TODO: handle pop
        return route.didPop(result);
      },
    );
  }
}
```

I basically listen to changes in my `selectedIdProvider` and force my `AppRouterDelegate` to reevaluate the navigation state by calling `notifyListeners`. Therefore I have to add `ChangeNotifier` to the class. It is very important to call `notifyListeners` on provider state changes and `read` the provider values where needed, instead of watching the provider directly inside the `build` method. Otherwise, the navigation will not work as expected. This would work exactly the same with a `StateNotifierProvider` instead of a `StateProvider`.

**Comment by echonox** [comment]
@daniellampl thanks for the suggestion, but I have another problem, the back button navigation doesn't work:

Router:
```
Router(
        routerDelegate: AppRouter(ref: ref),
        backButtonDispatcher: RootBackButtonDispatcher(),
 )
 ```

RouterDelegate:
```
class AppRouter extends RouterDelegate
    with ChangeNotifier, PopNavigatorRouterDelegateMixin {
  @override
  final GlobalKey<NavigatorState> navigatorKey;
  final WidgetRef ref;

  AppRouter({required this.ref}) : navigatorKey = GlobalKey<NavigatorState>() {
    ref.listen<AppState?>(appStateManager, (_, __) => notifyListeners());
  }

  @override
  Widget build(BuildContext context) {
    return Consumer(builder: (context, ref, child) {
      final state = ref.watch(appStateManager);
      return Navigator(
        key: navigatorKey,
        onPopPage: _onPopPage,
        pages: [
          if (state is Initial || state is LoggedOut) LoginScreen.page(),
          if (state is LoggedIn || state is TodoList) TodoListScreen.page(),
          if (state is FilteredTodoList) FilteredTodoListScreen.page(),
        ],
      );
    });
  }

  @override
  Future<void> setNewRoutePath(configuration) async => null;

  bool _onPopPage(Route route, result) {
    if (!route.didPop(result)) return false;

    if (route.settings.name == 'filteredtodolist') {}

    return true;
  }
}
```

StateNotifier:
```
final appStateManager =
    StateNotifierProvider.autoDispose<AppStateNotifier, AppState>((ref) {
  return AppStateNotifier();
});

class AppStateNotifier extends StateNotifier<AppState> {
  AppStateNotifier() : super(Initial());

  void login() => state = LoggedIn();
  void logout() => state = LoggedOut();
  void todolist() => state = TodoList();
  void filter() => state = FilteredTodoList();
}
```

Still the ```RootBackButtonDispatcher``` doesn't work, any suggestions?

**Comment by PSanJuanS** [comment]
@echonox I have a Navigator 2.0 steup quite similar to yours with backbutton working.
First , i don't have any listener in the constructor (your `ref.listen<AppState?>(appStateManager, (_, __) => notifyListeners());`) and works well , the Navigator gets updated everytime appState changes.

I think your problem is with the implementation of `_onPopPage,` you are not managing your state there, so your app state does not change with the backButton press, and the navigator does not update the pages.

try with:

```
 bool _onPopPage(Route route, result, AppStateNotifier notifier) {
    if (!route.didPop(result)) return false;

    if (route.settings.name == 'filteredtodolist') {
    notifier.todolist();
}

    return true;
  }
```
And edit this:

`onPopPage: (route,result){ _onPopPage(route,result,ref.watch(appStateManager.notifier))},`

This will trigger your todolist when you press back from your filteredtodolist.

Furthermore your page stack always will contain only one page due your disjunct states, i don't know if that is what you expect.

**Comment by ElteHupkes** [comment]
I'm also trying, and so far failing, to figure out how to do this properly. The main issue, it seems to me, is that there are two conflicting mechanisms at work here. The `Router` will use `RouterDelegate`'s `Listenable` to decide when to (a) call `build` and (b) update the state (i.e. call `RouteInformationParser<T>.restoreRouteInformation()`). If we `ref.watch()` for the app state inside `RouterDelegate.build`, that will trigger a rebuild in and of itself when the state changes. Thus, so far I see two equally bad options:

- Use `ref.watch(...)` inside `RouterDelegate.build()`, and don't call `notifyListeners()` anywhere. The side effect of this is that `restoreRouteInformation` is never called. This may not be very visible unless you're running a web app (where you'll notice the URL not changing) but it essentially kills state restoration.
- Somehow hook up `notifyListeners()` with `ref.listen(...)`. Now this is very icky to begin with, because from what I can tell you're really only supposed to call `ref.listen()` from within a `Consumer` widget's `build` method, and the constructor of your `RouterDelegate` is, well, not that. In order to get a correct `ref` instance, it seems to me you should be reconstructing your `RouterDelegate` every time, which means reconstructing the `Router` every time a route changes, which seems very wasteful to me. Now let's assume you get this wired up avoiding these issues, you now have the issue that the `ref.watch()` you're using inside the `build` that creates the `Navigator` will trigger a `rebuild`, but so will `notifyListeners()`, so now for every change in routing state, you're rebuilding the routing stack twice in a row. I suppose you could change `ref.watch` to `ref.read`, but that only works if the `RouterDelegate` is indeed recreated with a new `ref` every time, which is what you were probably trying to avoid to begin with.

In summary, none of this will work. I think personally I'm going to stop fighting the system, and use `Listenable` / `ChangeNotifier` for my `AppStateManager`, rather than a Riverpod `StateNotifier`. Of course the `AppStateManager` itself can still be provided through Riverpod.

**Comment by PSanJuanS** [comment]
@ElteHupkes Im not using route informations by now, so for me using the watch inside the build method is enough. Good to know that issue in advance ,because I have implementing a web version (that will need updated route info) in my  future roadmap. I will come around here if I figure it out once i get hands on that. Or if I surrender and have to change my navigationState structure.

**Comment by ElteHupkes** [comment]
@PSanJuanS So far it's only a minor surrender, I currently I plan to set it up like this (I have a mock app where this works fine):

- `AppStateManager with ChangeNotifier` calls `notifyListeners()` when the app state changes
- `final appStateManager = Provider((_) => AppStateManager())` is a simple provider for it. I wrap `MaterialApp.router()` in a `Consumer` widget that passes `ref.watch(appStateManager)` to my `AppRouter extends RouterDelegate`. This will only rebuild the entire router if the actual state manager changes, which won't happen.
- ...turtles all the way down from there.

So I'm basically sandwiching the `Listenable` state manager in the middle, the rest keeps working with Riverpod.

**Comment by PavelPZ** [comment]
I solved this problem as follows (my navigation status is affected by more StateProvider's):

1. my singleton RiverpodNavigator has a list of providers to listen to:
```
class RiverpodNavigator {
  RiverpodNavigator(Ref ref, List<AlwaysAliveProviderListenable> dependsOn>) {
    for (final depend in dependsOn) ref.listen<dynamic>(depend, (previous, next) => 
      onNavigationStateChanged()));
  }
  void onNavigationStateChanged() {
    routerDelegate.currentConfiguration = ...new navigation state, computed from all listened providers...;
    routerDelegate.notifyListeners();
  }
  RiverpodRouterDelegate routerDelegate = RiverpodRouterDelegate();
}
```

2. RiverpodNavigator is available from riverpodNavigatorProvider:
```
  final riverpodNavigatorProvider = Provider<RiverpodNavigator>(_(ref) => RiverpodNavigator (ref, [...]));
```

I prepared a working example here: [riverpod_navigator_example](https://github.com/PavelPZ/riverpod_navigator/tree/main/examples/riverpod_navigator_example);

**Comment by FXschwartz** [comment]
@PavelPZ Your link to the riverpod_navigator_example goes to a 404 page on Github. Do you still have that example?

**Comment by PavelPZ** [comment]
Link above fixed (I moved example to DartPad).

However, for some cases (where redirects or route guards are required) a different approach is necessary, see the [login-flow example](https://dartpad.dev/?id=a1df9d6b18e5e76b08d919b29fba6239).

I am preparing a medium article that explains it.

**Comment by PavelPZ** [comment]
And for my new project where I need async navigation, I'm experimenting with [riverpod_navigator package](https://github.com/PavelPZ/riverpod_navigator).

**Comment by PavelPZ** [comment]
The medium article about Riverpod & Navigation is here: [Simple Flutter navigation with Riverpod](https://medium.com/@pavelzika/simple-flutter-navigation-with-riverpod-c55b627da8a6).

**Comment by rrousselGit** [comment]
Closing in favor of #1023

**Comment by crizant** [comment]
I might be late for the party, but I just published a package for this purpose:
https://pub.dev/packages/riverpod_consumer_router_delegate

You guys may have a try a give me some feedback.

**Comment by gazialankus** [comment]
I'm even later to the party. Thank you @crizant for the package and your response in your repo. Upon further inspection I decided to go with a lighter route. 

@ElteHupkes, your note was immensely helpful and saved me from using `Consumer`/`ref.watch()` inside `RouterDelegate`'s `build()` method. One addition though to this point of yours about `ref.listen()`:

> * Somehow hook up `notifyListeners()` with `ref.listen(...)`. Now this is very icky to begin with, because from what I can tell you're really only supposed to call `ref.listen()` from within a `Consumer` widget's `build` method...

Exactly! For this we have `ref.listenManual()`, which I'm surprised nobody here mentioned. Maybe it got added later on, not sure. It's safe to call `listenManual()` in `initState()`, which is where I'm creating my `RouterDelegate`. So the `RouterDelegate``s constructor gets the `ref`, does `listenManual()` to listen to whatever is in Riverpod and copies them to variables in `RouterDelegate`, followed by a `notifyListeners()`. No reason this could not be a `state =` with `StateNotifier`. For the communication in the other direction (`RouterDelegate` to Riverpod), I can use `ref.read()` freely to access notifiers in Riverpod and call their methods. 

Anyone see any issues with this approach?
